Avoid using cert name in Secret / Order labels

srvaroa opened the pull request:
**What this PR does / why we need it**:

In the previous version, each Secret and Order would get a label with
the full name of the owning Certificate.  Because labels are limited to
63 chars, certificate names beyond that size would cause a number of
problems (e.g. certificates not getting renewed.)

For Secrets, the Certificate name is already being stored as an
annotation, so this patch removes the label altogether.  The annotation
may be used to retrieve the Certificate name from the secret when
needed.  This change is applied at
pkg/cainjector/controller:OwningCertForSecret as part of the patch.

For Orders, the Certificate is linked via an OwnerReference, which can
be used to filter.

**Which issue this PR fixes** 

fixes #1460

**Release note**:

```release-note
Fix issue with Certificate names being too long to use as labels.  There
is no label with the Certificate name being populated on Secrets or
Orders anymore, users should instead rely on the
"certmanager.k8s.io/certificate-name" annotation.
```

srvaroa commented:
@munnerz @cheukwing this is an alternative attempt to fix the issue following the conversation in https://github.com/jetstack/cert-manager/pull/1822, please let me know if it goes in the right direction.

munnerz commented on pkg/issuer/acme/issue.go:
```diff
@@ -218,13 +217,7 @@ func (a *Acme) Issue(ctx context.Context, crt *v1alpha1.Certificate) (*issuer.Is
-	selector := labels.NewSelector().Add(*req)
+	selector := labels.NewSelector()
 
 	existingOrders, err := a.orderLister.Orders(crt.Namespace).List(selector)
```
I think we can just use `selector.Everything()` here instead 😄

srvaroa replied:
I tried that but I believe the Lister gets a labels.Selector, which doesn't provide Everything (https://github.com/jetstack/cert-manager/blob/master/pkg/client/listers/certmanager/v1alpha1/order.go#L48) - I may very well be misreading though!

munnerz commented:
Also ref #2041 which makes a similar change in the acmeorders controller so that we no longer set an `order-name` label on Challenge resources.

srvaroa commented:
/retest

srvaroa commented on test/e2e/suite/issuers/acme/certificate/webhook.go:
```diff
@@ -179,19 +180,17 @@ var _ = framework.CertManagerDescribe("ACME webhook DNS provider", func() {
-					l, err := f.CertManagerClientSet.CertmanagerV1alpha1().Challenges(f.Namespace.Name).List(metav1.ListOptions{
-						LabelSelector: "acme.cert-manager.io/order-name=" + order.Name,
-					})
+					l, err := listOwnedChallenges(f.CertManagerClientSet, order)
```
@munnerz I copied this straight from your #2041

srvaroa commented on pkg/issuer/acme/issue.go:
```diff
@@ -426,7 +417,6 @@ func orderLabels(crt *v1alpha1.Certificate) map[string]string {
 	for k, v := range crt.Labels {
 		lbls[k] = v
 	}
-	lbls[certificateNameLabelKey] = crt.Name
```
@munnerz I was wondering whether to update `buildOrder` to include the annotation with the secret name, since given the change there is no trace of the cert name in the order anymore. But it seems to be unnecessary so far.

If you prefer to add it, I was thinking of sticking 
```go
Annotations:          orderAnnotations(crt),
```

with 

```go
func orderAnnotations(crt *v1alpha1.Certificate) map[string]string {
       return map[string]string{
               v1alpha1.CertificateNameKey: crt.Name,
       }
}
```

at L410. What do you think?

srvaroa commented:
Thanks for taking a look @munnerz 

> This is definitely the direction we'd prefer to go in, I think we just need to be aware of how we handle cleaning up existing resources when a user upgrades (i.e. because they are coming from a version that did not set the annotation).

Makes sense. I've addressed this here at https://github.com/jetstack/cert-manager/pull/2042/files#diff-d14e419095685c0725dd0282d442ae7dR458, ensuring that a) the annotation exists, and b) labels are cleaned up. Would that be correct? Let me know if I'm missing doing other shadow updates.

Thanks again!

srvaroa commented:
/retest

munnerz commented:
Just a thought - the ACME Order creation/management implementation in `pkg/issuer/acme` that you have modified is being surpassed by the new CertificateRequest based implementation that you can see here: https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificaterequests/acme/acme.go

Based on my understanding, this code does *not* use labels, which means it isn't subject to this kind of issue.

Given we're going to be switching the default implementation to this in the next release, I'm inclined to say we should not modify this behaviour at all in the old implementation given it will theoretically never land in a release (this code is due to be removed in the next week or so)

srvaroa commented:
@munnerz thanks, that makes sense. I removed that change from pkg/issuer/acme.  I also reverted a related change in e2e tests that ignored labels in order searches.

munnerz commented:
After a lot of recent changes in master, including rewriting the Orders controller and removing old code from the `certificates` controller, I think this PR is now obsolete 🙄 I think there is no longer anywhere that we set a label with the name of a resource, and we've now completely moved to either hashing the name of the resource, limiting it to 64 characters and using that, or otherwise just setting an annotation.

The v0.11.0-alpha.0 release will be cut in the next hour - if you can give this a go with a similar set of tests (i.e. certificates with really long names), we can double check it *is* fixed and if so, this PR can probably be closed 😄

srvaroa commented:
Thanks @munnerz! We can test with the next alpha release, although we will probably have to rebase #1338 first since we need that one to run in our environments. I'll get that moving.

munnerz commented:
I'm going to close this now as it's been superseded 😄 thanks for your work though, and look forward to working to help facilitate your requirements in #1338
